// @integration-app/sdk/udm/campaigns.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typebox_1 = require("@sinclair/typebox");
const FIELDS_SCHEMA = typebox_1.Type.Partial(typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    cost: typebox_1.Type.Number(),
    description: typebox_1.Type.String(),
    source: typebox_1.Type.String(),
    stage: typebox_1.Type.String(),
    type: typebox_1.Type.String(),
    ownerId: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    sentCount: typebox_1.Type.Number(),
    replyCount: typebox_1.Type.Number(),
    openCount: typebox_1.Type.Number(),
    clickCount: typebox_1.Type.Number(),
    startTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    endTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    createdTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    createdBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    updatedTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    updatedBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    lastActivityTime: typebox_1.Type.String({
        format: 'date-time',
    }),
}));
const MODIFIABLE_FIELDS = [
    'name',
    'description',
    'ownerId',
    'startTime',
    'endTime',
    'cost',
    'source',
    'stage',
    'type',
];
const udm = {
    singularName: 'campaign',
    pluralName: 'campaigns',
    fieldsSchema: FIELDS_SCHEMA,
    find: {
        queryFields: ['name'],
    },
    create: {
        fields: MODIFIABLE_FIELDS,
    },
    update: {
        fields: MODIFIABLE_FIELDS,
    },
};
exports.default = udm;
