// @integration-app/sdk/udm/companies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typebox_1 = require("@sinclair/typebox");
const base_1 = require("./base");
const FIELDS_SCHEMA = typebox_1.Type.Partial(typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    websiteUrl: typebox_1.Type.String(),
    phones: typebox_1.Type.Array(base_1.PhoneEmailSchema),
    primaryPhone: typebox_1.Type.String(),
    description: typebox_1.Type.String(),
    currency: typebox_1.Type.String(),
    industry: typebox_1.Type.String(),
    ownerId: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    primaryAddress: base_1.AddressSchema,
    addresses: typebox_1.Type.Array(base_1.AddressSchema),
    numberOfEmployees: typebox_1.Type.Number(),
    createdTime: typebox_1.Type.String({ format: 'date-time' }),
    createdBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    updatedTime: typebox_1.Type.String({ format: 'date-time' }),
    updatedBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    lastActivityTime: typebox_1.Type.String({ format: 'date-time' }),
}));
const MODIFIABLE_FIELDS = [
    'addresses',
    'currency',
    'description',
    'industry',
    'name',
    'numberOfEmployees',
    'ownerId',
    'primaryAddress',
    'primaryPhone',
    'websiteUrl',
];
const udm = {
    singularName: 'company',
    pluralName: 'companies',
    fieldsSchema: FIELDS_SCHEMA,
    find: {
        queryFields: ['name', 'websiteUrl'],
    },
    create: {
        fields: MODIFIABLE_FIELDS,
    },
    update: {
        fields: MODIFIABLE_FIELDS,
    },
    normalizeFields,
};
exports.default = udm;
function normalizeFields(fields) {
    const result = JSON.parse(JSON.stringify(fields));
    if (result.primaryPhone) {
        result.primaryPhone = (0, base_1.normalizePhone)(result.primaryPhone);
    }
    if (result.phones) {
        result.phones.forEach((phone) => {
            phone.value = (0, base_1.normalizePhone)(phone.value);
        });
    }
    return result;
}
