// @integration-app/sdk/udm/contacts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typebox_1 = require("@sinclair/typebox");
const base_1 = require("./base");
const FIELDS_SCHEMA = typebox_1.Type.Partial(typebox_1.Type.Object({
    fullName: typebox_1.Type.String(),
    firstName: typebox_1.Type.String(),
    lastName: typebox_1.Type.String(),
    primaryEmail: typebox_1.Type.String(),
    emails: typebox_1.Type.Array(base_1.PhoneEmailSchema),
    primaryPhone: typebox_1.Type.String(),
    phones: typebox_1.Type.Array(base_1.PhoneEmailSchema),
    primaryAddress: base_1.AddressSchema,
    addresses: typebox_1.Type.Array(base_1.AddressSchema),
    stage: typebox_1.Type.String(),
    companyName: typebox_1.Type.String(),
    companyId: typebox_1.Type.String({
        referenceUdm: 'companies',
    }),
    ownerId: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    jobTitle: typebox_1.Type.String(),
    source: typebox_1.Type.String(),
    createdTime: typebox_1.Type.String({ format: 'date-time' }),
    createdBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    updatedTime: typebox_1.Type.String({ format: 'date-time' }),
    updatedBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    lastActivityTime: typebox_1.Type.String({ format: 'date-time' }),
}));
const MODIFIABLE_FIELDS = [
    'fullName',
    'firstName',
    'lastName',
    'primaryEmail',
    'primaryPhone',
    'primaryAddress',
    'companyId',
    'companyName',
    'jobTitle',
    'source',
    'stage',
    'ownerId',
    'emails',
    'phones',
    'addresses',
];
const udm = {
    singularName: 'contact',
    pluralName: 'contacts',
    fieldsSchema: FIELDS_SCHEMA,
    find: {
        queryFields: ['fullName', 'primaryEmail', 'primaryPhone'],
    },
    create: {
        fields: MODIFIABLE_FIELDS,
    },
    update: {
        fields: MODIFIABLE_FIELDS,
    },
    normalizeFields,
    populateFields: base_1.populateContactInfo,
};
exports.default = udm;
function normalizeFields(fields) {
    const result = JSON.parse(JSON.stringify(fields));
    if (result.primaryPhone) {
        result.primaryPhone = (0, base_1.normalizePhone)(result.primaryPhone);
    }
    if (result.phones) {
        result.phones.forEach((phone) => {
            if (phone && typeof phone === 'object') {
                phone.value = (0, base_1.normalizePhone)(phone.value);
            }
        });
    }
    return result;
}
