// @integration-app/sdk/udm/deals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typebox_1 = require("@sinclair/typebox");
const FIELDS_SCHEMA = typebox_1.Type.Partial(typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    amount: typebox_1.Type.Number(),
    currency: typebox_1.Type.String(),
    companyId: typebox_1.Type.String({
        referenceUdm: 'companies',
    }),
    contactId: typebox_1.Type.String({
        referenceUdm: 'contacts',
    }),
    campaignId: typebox_1.Type.String({
        referenceUdm: 'campaigns',
    }),
    source: typebox_1.Type.String(),
    description: typebox_1.Type.String(),
    stage: typebox_1.Type.String(),
    status: typebox_1.Type.String({
        enum: ['open', 'won', 'lost'],
    }),
    type: typebox_1.Type.String(),
    ownerId: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    closeTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    createdTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    createdBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    updatedTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    updatedBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    lastActivityTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    probability: typebox_1.Type.Number(),
}));
const MODIFIABLE_FIELDS = [
    'amount',
    'currency',
    'companyId',
    'contactId',
    'campaignId',
    'source',
    'description',
    'stage',
    'status',
    'type',
    'ownerId',
    'closeTime',
];
const udm = {
    singularName: 'deal',
    pluralName: 'deals',
    fieldsSchema: FIELDS_SCHEMA,
    find: {
        queryFields: ['name'],
    },
    create: {
        fields: MODIFIABLE_FIELDS,
    },
    update: {
        fields: MODIFIABLE_FIELDS,
    },
};
exports.default = udm;
