// @integration-app/sdk/udm/deal-products.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typebox_1 = require("@sinclair/typebox");
const FIELDS_SCHEMA = typebox_1.Type.Partial(typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    quantity: typebox_1.Type.Number(),
    price: typebox_1.Type.Number(),
    dealId: typebox_1.Type.String({
        referenceUdm: 'deals',
    }),
    productId: typebox_1.Type.String({
        referenceUdm: 'products',
    }),
    isActive: typebox_1.Type.Boolean(),
    createdTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    createdBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    updatedTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    updatedBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
}));
const MODIFIABLE_FIELDS = ['name', 'quantity', 'price'];
const CREATABLE_FIELDS = ['name', 'quantity', 'price', 'dealId', 'productId'];
const udm = {
    singularName: 'deal-product',
    pluralName: 'deal-products',
    fieldsSchema: FIELDS_SCHEMA,
    find: {
        queryFields: ['dealId', 'productId', 'name'],
    },
    create: {
        fields: CREATABLE_FIELDS,
    },
    update: {
        fields: MODIFIABLE_FIELDS,
    },
};
exports.default = udm;
