// @integration-app/sdk/udm/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNIFIED_DATA_MODELS = exports.UDM = void 0;
const tslib_1 = require("tslib");
const activities_1 = tslib_1.__importDefault(require("./activities"));
const campaigns_1 = tslib_1.__importDefault(require("./campaigns"));
const companies_1 = tslib_1.__importDefault(require("./companies"));
const contacts_1 = tslib_1.__importDefault(require("./contacts"));
const deal_products_1 = tslib_1.__importDefault(require("./deal-products"));
const deals_1 = tslib_1.__importDefault(require("./deals"));
const products_1 = tslib_1.__importDefault(require("./products"));
const users_1 = tslib_1.__importDefault(require("./users"));
var UDM;
(function (UDM) {
    UDM["ACTIVITIES"] = "activities";
    UDM["CAMPAIGNS"] = "campaigns";
    UDM["COMPANIES"] = "companies";
    UDM["CONTACTS"] = "contacts";
    UDM["DEALS"] = "deals";
    UDM["DEAL_PRODUCTS"] = "deal-products";
    UDM["PRODUCTS"] = "products";
    UDM["USERS"] = "users";
})(UDM = exports.UDM || (exports.UDM = {}));
exports.UNIFIED_DATA_MODELS = {
    activities: activities_1.default,
    campaigns: campaigns_1.default,
    companies: companies_1.default,
    contacts: contacts_1.default,
    deals: deals_1.default,
    'deal-products': deal_products_1.default,
    products: products_1.default,
    users: users_1.default,
};
