// @integration-app/sdk/udm/products.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typebox_1 = require("@sinclair/typebox");
const FIELDS_SCHEMA = typebox_1.Type.Partial(typebox_1.Type.Object({
    name: typebox_1.Type.String(),
    description: typebox_1.Type.String(),
    code: typebox_1.Type.String(),
    type: typebox_1.Type.String(),
    isActive: typebox_1.Type.Boolean(),
    createdTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    createdBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
    updatedTime: typebox_1.Type.String({
        format: 'date-time',
    }),
    updatedBy: typebox_1.Type.String({
        referenceUdm: 'users',
    }),
}));
const MODIFIABLE_FIELDS = ['name', 'description', 'code', 'type', 'isActive'];
const udm = {
    singularName: 'product',
    pluralName: 'products',
    fieldsSchema: FIELDS_SCHEMA,
    find: {
        queryFields: ['name'],
    },
    create: {
        fields: MODIFIABLE_FIELDS,
    },
    update: {
        fields: MODIFIABLE_FIELDS,
    },
};
exports.default = udm;
