// @integration-app/sdk/src/udm/campaigns.ts
import { Static, Type } from '@sinclair/typebox'
import { UnifiedDataModel } from './base'

const FIELDS_SCHEMA = Type.Partial(
  Type.Object({
    name: Type.String(),
    cost: Type.Number(),
    description: Type.String(),
    source: Type.String(),
    stage: Type.String(),
    type: Type.String(),
    ownerId: Type.String({
      referenceUdm: 'users',
    }),
    sentCount: Type.Number(),
    replyCount: Type.Number(),
    openCount: Type.Number(),
    clickCount: Type.Number(),
    startTime: Type.String({
      format: 'date-time',
    }),
    endTime: Type.String({
      format: 'date-time',
    }),
    createdTime: Type.String({
      format: 'date-time',
    }),
    createdBy: Type.String({
      referenceUdm: 'users',
    }),
    updatedTime: Type.String({
      format: 'date-time',
    }),
    updatedBy: Type.String({
      referenceUdm: 'users',
    }),
    lastActivityTime: Type.String({
      format: 'date-time',
    }),
  }),
)

export type UnifiedCampaignFields = Static<typeof FIELDS_SCHEMA>

const MODIFIABLE_FIELDS = [
  'name',
  'description',
  'ownerId',
  'startTime',
  'endTime',
  'cost',
  'source',
  'stage',
  'type',
]

const udm: UnifiedDataModel = {
  singularName: 'campaign',
  pluralName: 'campaigns',
  fieldsSchema: FIELDS_SCHEMA,
  find: {
    queryFields: ['name'],
  },
  create: {
    fields: MODIFIABLE_FIELDS,
  },
  update: {
    fields: MODIFIABLE_FIELDS,
  },
}

export default udm
