// @integration-app/sdk/src/udm/contacts.ts
import { Static, Type } from '@sinclair/typebox'
import {
  AddressSchema,
  normalizePhone,
  PhoneEmailSchema,
  populateContactInfo,
  UnifiedDataModel,
} from './base'

const FIELDS_SCHEMA = Type.Partial(
  Type.Object({
    fullName: Type.String(),
    firstName: Type.String(),
    lastName: Type.String(),
    primaryEmail: Type.String(),
    emails: Type.Array(PhoneEmailSchema),
    primaryPhone: Type.String(),
    phones: Type.Array(PhoneEmailSchema),
    primaryAddress: AddressSchema,
    addresses: Type.Array(AddressSchema),
    stage: Type.String(),
    companyName: Type.String(),
    companyId: Type.String({
      referenceUdm: 'companies',
    }),
    ownerId: Type.String({
      referenceUdm: 'users',
    }),
    jobTitle: Type.String(),
    source: Type.String(),
    createdTime: Type.String({ format: 'date-time' }),
    createdBy: Type.String({
      referenceUdm: 'users',
    }),
    updatedTime: Type.String({ format: 'date-time' }),
    updatedBy: Type.String({
      referenceUdm: 'users',
    }),
    lastActivityTime: Type.String({ format: 'date-time' }),
  }),
)

const MODIFIABLE_FIELDS = [
  'fullName',
  'firstName',
  'lastName',
  'primaryEmail',
  'primaryPhone',
  'primaryAddress',
  'companyId',
  'companyName',
  'jobTitle',
  'source',
  'stage',
  'ownerId',
  'emails',
  'phones',
  'addresses',
]

export type UnifiedContactFields = Static<typeof FIELDS_SCHEMA>

const udm: UnifiedDataModel = {
  singularName: 'contact',
  pluralName: 'contacts',
  fieldsSchema: FIELDS_SCHEMA,
  find: {
    queryFields: ['fullName', 'primaryEmail', 'primaryPhone'],
  },
  create: {
    fields: MODIFIABLE_FIELDS,
  },
  update: {
    fields: MODIFIABLE_FIELDS,
  },
  normalizeFields,
  populateFields: populateContactInfo,
}

export default udm

function normalizeFields(fields: UnifiedContactFields): UnifiedContactFields {
  const result = JSON.parse(JSON.stringify(fields))
  if (result.primaryPhone) {
    result.primaryPhone = normalizePhone(result.primaryPhone)
  }
  if (result.phones) {
    result.phones.forEach((phone) => {
      if (phone && typeof phone === 'object') {
        phone.value = normalizePhone(phone.value)
      }
    })
  }
  return result
}
