// @integration-app/sdk/src/app-data-schemas/accessors.ts
import { AppDataSchema, AppDataSchemaInstance } from '.'
import { IntegrationAppClient } from '..'
import {
  ElementAccessor,
  ElementInstanceAccessor,
  ElementInstanceListAccessor,
  ElementListAccessor,
} from '../accessors'
import {
  AppDataSchemaInstanceSelector,
  CreateAppDataSchemaRequest,
  FindAppDataSchemaInstancesQuery,
  FindAppDataSchemasQuery,
  IAppDataSchemaInstanceUpdate,
  UpdateAppDataSchemaRequest,
} from './api'

export class AppDataSchemasAccessor extends ElementListAccessor<
  AppDataSchema,
  FindAppDataSchemasQuery,
  CreateAppDataSchemaRequest
> {
  constructor(client: IntegrationAppClient) {
    super(client, 'app-data-schemas')
  }
}

export class AppDataSchemaAccessor extends ElementAccessor<
  AppDataSchema,
  CreateAppDataSchemaRequest,
  UpdateAppDataSchemaRequest
> {
  constructor(client: IntegrationAppClient, idOrKey: string) {
    super({
      client,
      path: 'app-data-schemas',
      idOrKey,
    })
  }
}

export class AppDataSchemaInstancesAccessor extends ElementInstanceListAccessor<
  AppDataSchemaInstance,
  FindAppDataSchemaInstancesQuery
> {
  constructor(client: IntegrationAppClient) {
    super(client, 'app-data-schema-instances')
  }
}

export class AppDataSchemaInstanceAccessor extends ElementInstanceAccessor<
  AppDataSchemaInstance,
  AppDataSchemaInstanceSelector,
  IAppDataSchemaInstanceUpdate,
  IAppDataSchemaInstanceUpdate
> {
  constructor(
    client: IntegrationAppClient,
    selector: string | AppDataSchemaInstanceSelector,
  ) {
    super({
      client,
      instancePath: 'app-data-schema-instance',
      selector,
    })
  }
}
