// @integration-app/sdk/src/connections/index.ts
import { User } from '../users'
import { ConnectionMode, Integration } from '../integrations'
import { DataLocation } from '../data-locations'

export class Connection {
  /**
   * Unique connection id
   */
  id: string

  /**
   * User-readable name of the connection
   */
  name: string

  userId: string

  user?: User

  isDefunct?: boolean

  /**
   * Id of an integration the connection is created with
   */
  integrationId: string

  integration?: Integration

  createdAt: string

  updatedAt: string

  archivedAt?: string
}

export interface ConnectionSpec {
  name: string

  logoUri: string

  auth: {
    authorizeUri: string
    tokenUri: string
  }

  ui?: {
    description: string
    schema: any
    helpUri?: string
    test: boolean
  }

  parametersSchema?: any

  connectionMode?: ConnectionMode

  data?: ConnectionDataSpec
}

export class ConnectionSpec {
  constructor(opts: ConnectionSpec) {
    Object.assign(this, opts)
  }
}

export interface ConnectionDataSpec {
  root: DataLocation

  [key: string]: DataLocation
}
