// @integration-app/sdk/app-data-collections/accessors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppDataCollectionInstanceAccessor = exports.AppDataCollectionInstancesAccessor = exports.AppDataCollectionAccessor = exports.AppDataCollectionsAccessor = void 0;
const accessors_1 = require("../accessors");
class AppDataCollectionsAccessor extends accessors_1.ElementListAccessor {
    constructor(client) {
        super(client, 'app-data-collections');
    }
}
exports.AppDataCollectionsAccessor = AppDataCollectionsAccessor;
class AppDataCollectionAccessor extends accessors_1.ElementAccessor {
    constructor(client, idOrKey) {
        super({
            client,
            path: 'app-data-collections',
            idOrKey,
        });
    }
}
exports.AppDataCollectionAccessor = AppDataCollectionAccessor;
class AppDataCollectionInstancesAccessor extends accessors_1.ElementInstanceListAccessor {
    constructor(client) {
        super(client, 'app-data-collection-instances');
    }
}
exports.AppDataCollectionInstancesAccessor = AppDataCollectionInstancesAccessor;
class AppDataCollectionInstanceAccessor extends accessors_1.ElementInstanceAccessor {
    constructor(client, selector) {
        super({
            client,
            instancePath: 'app-data-collection-instance',
            selector,
        });
    }
}
exports.AppDataCollectionInstanceAccessor = AppDataCollectionInstanceAccessor;
