// @integration-app/sdk/data-builder/context.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DataBuilderContext = void 0;
const errors_1 = require("../errors");
class DataBuilderArrayIterator {
}
class DataBuilderContext {
    constructor(variables) {
        this.variables = variables;
        this.currentArrayItems = {};
    }
    setCurrentArrayItem(locator, value) {
        this.currentArrayItems[this.getKeyByLocator(locator)] = value;
    }
    getCurrentArrayItem(locator) {
        const key = this.getKeyByLocator(locator);
        if (typeof this.currentArrayItems[key] !== 'undefined') {
            return this.currentArrayItems[key];
        }
        else {
            throw new errors_1.BadRequestError(`Attempting to access current array item with non-existing locator:\n${JSON.stringify(locator)}`);
        }
    }
    unsetCurrentArrayItem(locator) {
        delete this.currentArrayItems[this.getKeyByLocator(locator)];
    }
    getKeyByLocator(locator) {
        return JSON.stringify(locator);
    }
}
exports.DataBuilderContext = DataBuilderContext;
