// @integration-app/sdk/flows/nodes/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FLOW_NODE_SPECS = exports.FlowNodeType = exports.FlowNodeSpec = void 0;
var base_1 = require("./base");
Object.defineProperty(exports, "FlowNodeSpec", { enumerable: true, get: function () { return base_1.FlowNodeSpec; } });
var FlowNodeType;
(function (FlowNodeType) {
    FlowNodeType["ApiTrigger"] = "api-trigger";
    FlowNodeType["ScheduleTrigger"] = "schedule-trigger";
    FlowNodeType["AppEventTrigger"] = "app-event-trigger";
    FlowNodeType["DataRecordCreatedTrigger"] = "data-record-created-trigger";
    FlowNodeType["DataRecordUpdatedTrigger"] = "data-record-updated-trigger";
    FlowNodeType["DataRecordDeletedTrigger"] = "data-record-deleted-trigger";
    FlowNodeType["TransformData"] = "transform-data";
    FlowNodeType["ListDataRecords"] = "list-data-records";
    FlowNodeType["LookupDataRecord"] = "lookup-data-record";
    FlowNodeType["FindDataRecordById"] = "find-data-record-by-id";
    FlowNodeType["CreateDataRecord"] = "create-data-record";
    FlowNodeType["UpdateDataRecord"] = "update-data-record";
    FlowNodeType["DeleteDataRecord"] = "delete-data-record";
    FlowNodeType["HttpRequest"] = "http-request";
    FlowNodeType["Filter"] = "filter";
    FlowNodeType["CreateDataLink"] = "create-data-link";
    FlowNodeType["FindDataLink"] = "find-data-link";
    FlowNodeType["FindDataRecords"] = "find-data-records";
})(FlowNodeType = exports.FlowNodeType || (exports.FlowNodeType = {}));
exports.FLOW_NODE_SPECS = {
    [FlowNodeType.ApiTrigger]: require('./api-trigger').default,
    [FlowNodeType.ScheduleTrigger]: require('./schedule-trigger').default,
    [FlowNodeType.AppEventTrigger]: require('./app-event-trigger').default,
    [FlowNodeType.DataRecordCreatedTrigger]: require('./data-record-created-trigger').default,
    [FlowNodeType.DataRecordUpdatedTrigger]: require('./data-record-updated-trigger').default,
    [FlowNodeType.DataRecordDeletedTrigger]: require('./data-record-deleted-trigger').default,
    [FlowNodeType.ListDataRecords]: require('./list-data-records').default,
    [FlowNodeType.LookupDataRecord]: require('./lookup-data-record').default,
    [FlowNodeType.FindDataRecords]: require('./find-data-records').default,
    [FlowNodeType.FindDataRecordById]: require('./find-data-record-by-id')
        .default,
    [FlowNodeType.CreateDataRecord]: require('./create-data-record').default,
    [FlowNodeType.UpdateDataRecord]: require('./update-data-record').default,
    [FlowNodeType.DeleteDataRecord]: require('./delete-data-record').default,
    [FlowNodeType.TransformData]: require('./transform-data').default,
    [FlowNodeType.Filter]: require('./filter').default,
    [FlowNodeType.HttpRequest]: require('./http-request').default,
    [FlowNodeType.CreateDataLink]: require('./create-data-link').default,
    [FlowNodeType.FindDataLink]: require('./find-data-link').default,
};
