// @integration-app/sdk/data-locations/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeCollectionOperationPath = exports.makeDataLocationOperationPath = exports.SubscriptionType = exports.DataLocationType = void 0;
const tslib_1 = require("tslib");
const url_join_1 = tslib_1.__importDefault(require("url-join"));
var DataLocationType;
(function (DataLocationType) {
    DataLocationType["directory"] = "directory";
    DataLocationType["collection"] = "collection";
})(DataLocationType = exports.DataLocationType || (exports.DataLocationType = {}));
var SubscriptionType;
(function (SubscriptionType) {
    SubscriptionType["PUSH"] = "push";
    SubscriptionType["PULL"] = "pull";
})(SubscriptionType = exports.SubscriptionType || (exports.SubscriptionType = {}));
function makeDataLocationOperationPath(collectionPath, operation) {
    const [path, query] = (collectionPath !== null && collectionPath !== void 0 ? collectionPath : '').split('?');
    return (0, url_join_1.default)(path, operation, query ? `?${query}` : '');
}
exports.makeDataLocationOperationPath = makeDataLocationOperationPath;
exports.makeCollectionOperationPath = makeDataLocationOperationPath;
