// @integration-app/sdk/flows/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getOrderedNodeKeys = exports.getRootNodeKeys = exports.getUpstreamNodeKeys = exports.getParentNodeKeys = exports.getDownstreamNodeKeys = exports.getChildNodeKeys = exports.getFlowNode = exports.FlowInstanceNodeState = exports.FlowInstanceState = exports.FLOW_INSTANCE_CHANNEL_PREFIX = exports.FLOW_CHANNEL_PREFIX = void 0;
exports.FLOW_CHANNEL_PREFIX = 'private-flows-';
exports.FLOW_INSTANCE_CHANNEL_PREFIX = 'private-flow-instances-';
var FlowInstanceState;
(function (FlowInstanceState) {
    FlowInstanceState["SETTING_UP"] = "SETTING_UP";
    FlowInstanceState["SETUP_FAILED"] = "SETUP_FAILED";
    FlowInstanceState["READY"] = "READY";
})(FlowInstanceState = exports.FlowInstanceState || (exports.FlowInstanceState = {}));
var FlowInstanceNodeState;
(function (FlowInstanceNodeState) {
    FlowInstanceNodeState["SETTING_UP"] = "SETTING_UP";
    FlowInstanceNodeState["SETUP_FAILED"] = "SETUP_FAILED";
    FlowInstanceNodeState["READY"] = "READY";
})(FlowInstanceNodeState = exports.FlowInstanceNodeState || (exports.FlowInstanceNodeState = {}));
var utils_1 = require("./utils");
Object.defineProperty(exports, "getFlowNode", { enumerable: true, get: function () { return utils_1.getFlowNode; } });
Object.defineProperty(exports, "getChildNodeKeys", { enumerable: true, get: function () { return utils_1.getChildNodeKeys; } });
Object.defineProperty(exports, "getDownstreamNodeKeys", { enumerable: true, get: function () { return utils_1.getDownstreamNodeKeys; } });
Object.defineProperty(exports, "getParentNodeKeys", { enumerable: true, get: function () { return utils_1.getParentNodeKeys; } });
Object.defineProperty(exports, "getUpstreamNodeKeys", { enumerable: true, get: function () { return utils_1.getUpstreamNodeKeys; } });
Object.defineProperty(exports, "getRootNodeKeys", { enumerable: true, get: function () { return utils_1.getRootNodeKeys; } });
Object.defineProperty(exports, "getOrderedNodeKeys", { enumerable: true, get: function () { return utils_1.getOrderedNodeKeys; } });
