// @integration-app/sdk/src/data-sources/index.ts
import { User } from '../users'
import { Connection } from '../connections'
import { Integration } from '../integrations'
import { ErrorData } from '../errors'

export interface DataSourceInstance {
  id: string

  uuid: string

  userId: string

  user?: User

  udm?: string

  connectionId: string

  connection?: Connection

  integrationId: string

  integration?: Integration

  instanceKey?: string

  /**
   * Path to the root of the data source (where user can choose the specific path from)
   */
  rootPath?: string

  /**
   * Currently selected path.
   */
  path?: string

  /**
   * Whether the data source instance was customized by the user.
   *
   * If it was customized - it won't automatically update when the underlying data changes.
   */
  isCustomized?: boolean

  error?: ErrorData

  subscribedTo?: {
    created?: boolean
    updated?: boolean
    deleted?: boolean
  }

  nextPullTimestamp?: number

  pullAttempts?: number
}

export interface DataSourceEvent {
  id: string

  userId: string

  user?: User

  dataSourceInstanceId: string

  dataSourceInstance?: DataSourceInstance

  integrationId: string

  integration?: Integration

  connectionId: string

  connection?: Connection

  event: any

  datetime: string

  launchedFlowRunIds: string[]
}
