// @integration-app/sdk/src/app-data-collections/accessors.ts
import { AppDataCollection, AppDataCollectionInstance } from '.'
import { IntegrationAppClient } from '..'
import {
  ElementAccessor,
  ElementInstanceAccessor,
  ElementInstanceListAccessor,
  ElementListAccessor,
} from '../accessors'
import {
  AppDataCollectionInstanceSelector,
  CreateAppDataCollectionRequest,
  FindAppDataCollectionInstancesQuery,
  FindAppDataCollectionsQuery,
  AppDataCollectionInstanceUpdate,
  UpdateAppDataCollectionRequest,
} from './api'

export class AppDataCollectionsAccessor extends ElementListAccessor<
  AppDataCollection,
  FindAppDataCollectionsQuery,
  CreateAppDataCollectionRequest
> {
  constructor(client: IntegrationAppClient) {
    super(client, 'app-data-collections')
  }
}

export class AppDataCollectionAccessor extends ElementAccessor<
  AppDataCollection,
  CreateAppDataCollectionRequest,
  UpdateAppDataCollectionRequest
> {
  constructor(client: IntegrationAppClient, idOrKey: string) {
    super({
      client,
      path: 'app-data-collections',
      idOrKey,
    })
  }
}

export class AppDataCollectionInstancesAccessor extends ElementInstanceListAccessor<
  AppDataCollectionInstance,
  FindAppDataCollectionInstancesQuery
> {
  constructor(client: IntegrationAppClient) {
    super(client, 'app-data-collection-instances')
  }
}

export class AppDataCollectionInstanceAccessor extends ElementInstanceAccessor<
  AppDataCollectionInstance,
  AppDataCollectionInstanceSelector,
  AppDataCollectionInstanceUpdate,
  AppDataCollectionInstanceUpdate
> {
  constructor(
    client: IntegrationAppClient,
    selector: string | AppDataCollectionInstanceSelector,
  ) {
    super({
      client,
      instancePath: 'app-data-collection-instance',
      selector,
    })
  }
}
